source/extensions/filters/network/zookeeper_proxy/zookeeper_utils.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+#include <cstdint>
+
+#include "envoy/common/platform.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/byte_order.h"
+#include "common/common/logger.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ZooKeeperProxy {
+
+/**
+ * Helper for extracting ZooKeeper data from a buffer.
+ *
+ * If at any point a peek is tried beyond max_len, an EnvoyException
+ * will be thrown. This is important to protect Envoy against malformed
+ * requests (e.g.: when the declared and actual length don't match).
+ *
+ * Note: ZooKeeper's protocol uses network byte ordering (big-endian).
+ */
+class BufferHelper : public Logger::Loggable<Logger::Id::filter> {
+public:
+  BufferHelper(uint32_t max_len) : max_len_(max_len) {}
+
+  int32_t peekInt32(Buffer::Instance& buffer, uint64_t& offset);
+  int64_t peekInt64(Buffer::Instance& buffer, uint64_t& offset);
+  std::string peekString(Buffer::Instance& buffer, uint64_t& offset);
+  bool peekBool(Buffer::Instance& buffer, uint64_t& offset);
+  void add(uint32_t size) { current_ += size; }
```

Comment:
It seems like current_ is only incremented here - what's the purpose of maintaining one internal to the buffer in addition to the external offset value?

Reply:
this one is used for the currently being read message, whereas the external offset refers to the buffer which may contain many messages.

Maintaining `current_` allows us to centralize checks about a message's length, ensuring we are never reading beyond `max_packet_bytes_`...